Availability from bvh

tfili opened the pull request:
We now don't use the availability list from layer.json. Instead we load it from the BVH terrain extension. Here is a Sandcastle example. The GTOPO terrain is stand alone and the Zion UT terrain is cutout terrain on top of GTOPO.

https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/index.html#c=tVZtb+I4EP4rVr9AJWoSJ06cLq2uR3tVqgLblu7tcpxOJjHgNrHZxLyu+t9vQqClqD1dpV1/AcYz88w880xCvY4uM64MaopcTtOrO8SjSOQ5Mhot9TRDUivE81yYvK9KHxxqhWMx5NPEnK2du/pRKHSCKmJ5NR5cRrIjr8L7VWi3ZZiH6pZGzdALHydfvzSvAgxO3+PLx8Lp8Vt6lfbOo0VracvWw8X8unuzbKXfTOcydDpNO213b8ft8xjuRotW2pLXzatJD5K1lnPJ//zDCh/0ot0NSev8nrZXZ/PhDY6GzZvv5mjs9mLX8H++XjysTOBPHdbknWC2WMyH/t3N7ywcxGHl01s93YlsJrKinbExk/y4Xo/F7IhPJI5K30inRWRfzXiGRkZPdFdkGZcFBUrMN1Ti8mNz9TnTMxmLrPqjrxCcaZYcoxfsW5ED2ZHAw0ynZwXdYVz1vMO+ejp8hlrBLH4Rkv8aaSbFfE3BDsiXta1aKUloamUATGSV2gbHvIZHx6+oeU6/0jrt6kJzcZOrpVbV1w1eLIwAOZ5sYW9FZLgaJWXBtzyWXOXVIxsHPvGJZfm2YxHbdUnNwp7rM5cxy3H8wPfBtHYDF+JQ4lPP9pmzdgtcSqntgsXzbLYpYN0fziOhBB4leiCwUHyQiGs5GhupRlCTyaZi7dxXd9BAxHMDhfE47mqdDHjWEmpa/euZkIUBFiqX3c7njmNVaqVZq1wk0BRcDacqMrKgAG1iirMpZJ/Pk1d8fir9n4DXGtoDbF+cIztFeoh6kLyG7rsACgs9KUw/rZodNe4U83dJ5gfYKZSASin876LeFtG7hGxp+GD+ntxLvGluG/NWGdssG9oinoqM42Gy7OrqDkAscpAUX2d53s4mzwx848pZa/1cjDIhCq3bBNuuV2jac2hAiFdDjoct6jCLENdxQcge2Aj1LexQ3wG9O5bLHO+w9gKpMwmLtYXcqaU4YwGLBRJ/rqXFzRgbvd03EmAvgPw2Za4X2D7dzVyciTTR+N3wI8fHlHmwdo5nW4y5zn58ppMEwi1sw1oGsNIeC4jveoy++D1tJlFM4emNQfTkr5xAwGhAHcv3CHOICxPwYSiO7VECxVoWpTXkMpdiBhOwGC1aID+Pf4Jdy3d9yijMlgAvHx2ABw1YLmEucRmhHqHvTgDmQy1mQS8WCaBn9h8T6KuD2kEjN8tEnJZ3v8l0ojNTvHaqGNeNSCcJB1Lrg2n0KAyO8ryIbtS3QY1YzpCMT/oHe2+W/gGKEvjrATfDaZLcyZXoH5w26uD/KizRa+468NJO+BJcykIaY/v0urzCGDfq8LOA3Y825bNpJ/O/

pjcozzi commented:
@tfili please submit an issue about backwards compatibility / deprecation as we discussed?

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -382,7 +415,15 @@ define([
+         * @constant
+         * @default 3
+         */
+        BVH: 3
```
Are we sure this shouldn't be spelled out like `WATER_MASK`?  Perhaps similar question for the extension spec if that is relevant.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -512,10 +554,34 @@ define([
+                    layer.bvhLoaded.addAvailableTileRange(level, x, y, x, y);
+                    recurseHeights(level, x, y, bvh, 0, maxLevel, provider.availability, layer.availability);
+                } else {
+                    console.log('Incorrect number of heights in tile Level: %d X: %d Y: %d', level, x, y);
```
Is `console.log` the right approach compared to an exception?

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -556,6 +622,42 @@ define([
+
+        if (!isNaN(buffer[index])) {
+            // Minimum height isn't a Nan, so the tile exists
+            // TODO: Make this more efficient
```
We generally don't merge PRs with `TODOs`.  Write an issue or just do it.  Is it a bottleneck in practice?

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -772,6 +880,26 @@ define([
+                }
+                //>>includeEnd('debug');
+
+                // returns true if we can request vertex normals from the server
```
Is this comment old?

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -798,6 +926,19 @@ define([
+         * @memberof CesiumTerrainProvider.prototype
+         * @type {Boolean}
+         */
+        requestBvh : {
```
Not sure that we should use the abbreviation in the public API.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -840,9 +1015,93 @@ define([
+
+        var promise = when.resolve();
+        if (level !== 0) {
+            var bvhLevels = layer.bvhLevels - 1;
```
Extract out into a helper function given the duplication with the code above?

kring commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -828,6 +969,40 @@ define([
+        var level = this.availability._maximumLevel;
+        this._tilingScheme.positionToTileXY(position, level, scratchTileXY);
+
+        return when(loadTileAvailability(this, scratchTileXY.x, scratchTileXY.y, level));
```
So every time you ask the terrain provider to load availability it will create a promise, and then, if availability isn't yet know for the maximum level at that position, it will do a non-throttled request for the closest BVH node to that level (generating 404s if it doesn't exist). I modifed the Terrain sandcastle example to use your Zion terrain and sample heights around Zion. Pressing the "Sample Most Detailed Everest Terrain" button generates thousands of requests all at once and locks up the browser. Here's the [Super Hacky Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=xVlbd9pIEv4rvXlBPsGN7pc4yVkbOwkZYyc2k0wS5+xppAY6FmoiNQY8x2/7z/aPbbVuSCA8ZDOzowcD6rpXddXX7S5N2HyKezzCAR2ReSiOfZ8myYDf0gi9QC26ejsZvvbZJXvb+/W+p12wXtKLriy/27N7t7PfPnTfehiIvgevbyXR7afp2+nnU3/ZX2ms/+1scT54v+pPP4nL1z3jsqtNLwZXk4vTANbGy/60z867b2efQVh/tWDk4yu1940vLwY9vX/6q3Vxf7wYvcf+qPv+uzicmJ8DU5B//Xb27V54ztxwu+TSu1suFyPn+v2J2xsGvdbRTdTd8umaxnc0lu5MhJglzzqdgN4dkhnDfkbr86nkvInuSIzuGY8GNI4JkxGI6ALlErOPfOldzO9YQGPl95sIwTOPw2dorfqKJnwe+xSPYj49ThIqeoFiOwc30cNBqemO0UVqV0XJh/Sd0sos6/JIgDIat9ooVyTq+tGzqsGl+E4HgU4UsvFEsGiMBEcinlPQm8rHiU8jischH1JMIzIM6XlB+iKlrEiZzxAJQW1EBLujhQFolluQZM7QMGSzhLOgtKzm11mxvBm/SjjiwYR+5HG4Q8SHq2J9dw4gHGmSIceLxQLfxYdiQheSRya5U/ndESykiYZVrHYco9NqZzL8mAZMSDFFDQREEORDNgVNVujDAPX/8+9fWmWkr0kU+CQRIcUkCAach0MS92k0V76UGVumAhvrBx3m7qHUM5QvF+bwKKEh9SX/aB75AlKtHBSlIJ88n5tV8QItKoE62qLfJ/+S/AH8rJTejzsCryOO6GgETiQ/71VeCpAlImq1kNbRn2PxooOKYPyVBle45RPT73OaiA8U6mx5weMpCRNQl+3agujh4O9MJQTmI3gR//+jkqrtk+R2d0Qa3dvVdX7eg61295gZOzvXz9ux3TQrhnxto1bOIjtSq6ljnWYz8rHG1efzSKAzmKGQir1NDjm/PRZ9kfM9vkHfkHCETvmU7h8RmBiCxGOYcvUpTWSjZiQylEPddD1bt7DqGrauW5rptNGhaXiGZ2rYhMdQdbONDNfVLcPCnmmZmmO4nlPbZqCIj0bJY4ps13KwqXqqrhqOZdptZGo6NnTXdA3L9kzDcttIB/3YskCJbruarupGw272yZTGBGfBUzIH27n+PyIfxCRKRtA6lNzEPhExW5q4d3oGmGrw6dEUQFmgVyDBZ4nP0QlZ/bmpcFTX1UzsWpZrQF5kInQIt+Vh01Jd0/JUR2bC00zVkJlwTN2Djx/MhKPapo5tW4XMqp5pW6DH0Gxbw5bmeJANzXFTRZalWhheeZ6jmYZh2H9nLuQ+ved8ejIXAqLdWmOifPdek+kspO94wmT0k3S5yMX2Ritys0deDFV1HBVDBhwdgm+4sB0s2zA9KF/N0GCjWC7EUPccV7N0rLu2LGBNK8O1R05AnIk9B7aHazuuZeuQE8czsGFqnuO5qqs5pgY6TM/DHlji2qpjQ9L0Usf+6fifU/FQC2mSxjtv1tfz9EykNOeiFu1yKqxnXDmC8MfX166Zpk6SdzonK5QfUNrFvA04TQAvCcSHyTymaEyhJYp4VbAM6YRFQVEVGF1TkcH7CUvQKCRjlCGABN4QgWsxqeKEgM7EZADFcjwGOYlYn3dK7F/hpJEAX2mCk3kyo1EAhRaJRNmIeUkV0ym/o8dhmGN7SQPhR4qMEAMd6hF8PN9R2zik0VhMjtDTp2yry8xyKmloI/cX9rXU2WQbDLxNiBFBrUBrK0TjCZXwCQv+ii1poGgH7Tp9acOzdbqh4GLBxzGZTZg/4GXxKwXxppAhC8MhJ3EAUjbsSc0GFMh8El7GbMykpuIIVHuPTy4Hg8t+e5s/ARrplIqdhlU2JWO52sK4k35POiPis5CJFR6zUavO8bAhISRDGjabnY+SvUOZVUaUjh9NnQk05RFPZsQvYUD1mfCY3csD8VZY3mys4C5s7LOrBhkzMCS8zPrVs8Z+pSsq9CfNbLQVstblIY/XqtOf+OT8uPtLAwOfixCO7z/Ek0z44oT4t+MYYFeQA94GumFJU4ivuiNfKSr22qj6xzl4VNA72B6yneyKjMRKBxvV0QzFm/tCMp/SSvOo99zsMPA6ZoESA9og0TikEhZes3taa7JjIPnIAjGBLmBqR/WFN2nJVVc6nbQzZ8PxnIAx84Cu+zMMgwlU6BUJwMVE0R3sSThYQsA6O4/Gj/O7NvYk0rQ2+OvT8apwUNkSfIi2vG9vWf8oUSnq6T6iGojWbbswfdPs9FqrcPlQg+muOzpACQ1gl2aaelvFtukA+nVVw3A8B8BcOyUDEt2wdMeyAYgZKRlADEsGzJYwza3rbm7yYNCXr0cbg2WVDZYVDJZ1HcgpsqpNkZJ+mdEvc/q0oCT58mCzs0nqcEfiQxrPFIoXEFEILaZEfi5RBynrKj1E2kEVLpYym2sxF5lA75hImRGP5ZdVITSv8F1SKyNyYxMX80kpnWmXNmxK2jGbZ/Nksp5pFZ6H2s6PqZjH0W7w+tB0cTaGusoAsNI6S1FM5Rrmjy872uvDiT+h/i0N1k1jj4uSnOdo571eg3kjvmUZvMplB/saVGHZx470mF4YkoW2OJqXFzVEoHPY5yHyWu2GM9ujO6vShlWsqmUfywtpMaFRgaNrKFlpvqJoIxg8O5BzuxFnFwqb7g/g1NrKeOpHpX2C1OcQoVMqCJOR3ghZY5zYCCn/yH0FpA6DPNjhJSZ3IJYMUwx1UOsgPljJwa6Qj5VWzV9pUGkPy3B/Mp/NYL/DC9moxISi7OxNg61791Z1/2U7bmP+/kCa989z1eqdOf/LE/6k/eR5IlYhfZnx/pNNZdzkvwAUQLWCAgv4l3SGc9hOAvuZmuedgul5wO4QC17cPNn4T8vNE+SHJElgZTQPQzkPb568fN4B+hpbyInESpdgbEhWkmSivTzPXmKMn3fg5zaXyEoTyGtv13d0jboqdwON67UIVSnyz/8C).

kring commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -512,10 +554,34 @@ define([
+                var bvh = new Float32Array(buffer, extensionPos, numberOfHeights);
+                extensionPos += Float32Array.BYTES_PER_ELEMENT * numberOfHeights;
+
+                var numberofIncludedLevels = findNumberOfLevels(numberOfHeights/2);
```
The iterative algorithm in `findNumberOfLevels` is pretty expensive to do on every tile, and I don't think it's necessary. For the error checking part, just calculate the number of tiles that should be in the BVH at this level, and see if the actual number of tiles is the same. For `recurseHeights`, you just need to adjust your ranges as you recurse. For the loaded tile, the BVH range is 0 through length-1. Then the number of BVH tiles in each child is n = (length - 1) / 4, the southwest child is 1 through n, southeast child is n + 1 through 2n, northwest is 2n + 1 through 3n, and northeast is 3n + 1 through 4n.

kring commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -840,9 +1015,93 @@ define([
-        return this._availability.isTileAvailable(level, x, y);
+        var result = loadTileAvailability(this, x, y, level);
+
+        // If we got a result, return it. If it was a promise return false for now.
```
In the `fill-tiles` branch, returning false here will cause the terrain engine to upsample the tile instead of loading it, and never look back. We should return some sort of sentinel value (null? unfortunately we're already using undefined) to indicate that we don't know yet, rather than returning false.

kring commented:
That's all I have. I still wonder if it would be wiser to use some kind of tile-local version of the rectangle ranges we're currently using in layer.json, rather than the BVH extension.

pjcozzi commented:
Thanks for the quick review, @kring!

> I still wonder if it would be wiser to use some kind of tile-local version of the rectangle ranges we're currently using in layer.json, rather than the BVH extension.

@tfili what do you think of this?  Sounds like there are some performance concerns with the current approach.

tfili commented:
@pjcozzi We have a metadata extension in the works and mostly ready to go. I'm reviewing that today. I can use that send down subtree available ranges, instead of using the BVH extension. That should address most of the performance issues @kring brings up. The good thing here is that most of the code will be reused and actually slightly simplified.

Thanks @kring for the review. I didn't realize the availability bloat with my approach. I have a better understanding of whats going on now.

pjcozzi commented:
Cool, good luck @tfili!

tfili commented:
@kring If you wouldn't mind taking another look it would be appreciated. I abandoned using BVH for availability and  instead store the ranges in the tiles every 10 levels. The only thing that has changed is the asset numbers. I also cache promises for availability requests so we don't make a bunch of the same ones.

The one issue with your branch is that we are returning false from `getTileDataAvailable`. My first thought is that we can just return a promise. Maybe down the road we use the promise, but for now it can just be the value used to signal a retry. I think it probably needs to be made in your branch, because it sounds like you current change would break `GoogleEarthEnterpriseTerrainProvider` as it currently returns false when it needs to load a parent tile to check availability.

Here is the example:

http://localhost:8080/Apps/Sandcastle/index.html#c=tVZtb+I4EP4rVr9AJWoSJ06cLq2uR3tVqgLblu7tcpxOJjHgNrHZxLyu+t9vQqClqD1dpV1/AcYz88w880xCvY4uM64MaopcTtOrO8SjSOQ5Mhot9TRDUivE81yYvK9KHxxqhWMx5NPEnK2du/pRKHSCKmJ5NR5cRrIjr8L7VWi3ZZiH6pZGzdALHydfvzSvAgxO3+PLx8Lp8Vt6lfbOo0VracvWw8X8unuzbKXfTOcydDpNO213b8ft8xjuRotW2pLXzatJD5K1lnPJ//zDCh/0ot0NSev8nrZXZ/PhDY6GzZvv5mjs9mLX8H++XjysTOBPHdbknWC2WMyH/t3N7ywcxGHl01s93YlsJrKinbExk/y4Xo/F7IhPJI5K30inRWRfzXiGRkZPdFdkGZcFBUrMN1Ti8mNz9TnTMxmLrPqjrxCcaZYcoxfsW5ED2ZHAw0ynZwXdYVz12WFfPR0+Q61gFr8IKXiNNJNivqZgB+TL2latlCQ0tTIAJrJKbYNjXsOj41fUPKdfaZ12daG5uMnVUqvq6wYvFkaAHE+2sLciMlyNkrLgWx5LrvLqkY0Dn/jEsnzbsYjtuqRmYc/1mcuY5Th+4PtgWruBC3Eo8aln+8xZuwUupdR2weJ5NtsUsO4P55FQAo8SPRBYKD5IxLUcjY1UI6jJZFOxdu6rO2gg4rmBwngcd7VOBjxrCTWt/vVMyMIAC5XLbudzx7EqtdKsVS4SaAquhlMVGVlQgDYxxdkUss/nySs+P5X+T8BrDe0Bti/OkZ0iPUQ9SF5D910AhYWeFKafVs2OGneK+bsk8wPsFEpApRT+d1Fvi+hdQrY0fDB/T+4l3jS3jXmrjG2WDW0RT0XG8TBZdnV1ByAWOUiKr7M8b2eTZwa+ceWstX4uRpkQhdZtgm3XKzTtOTQgxKshx8MWdZhFiOu4IGQPbIT6Fnao74DeHctljndYe4HUmYTF2kLu1FKcsYDFAok/19LiZoyN3u4bCbAXQH6bMtcLbJ/uZi7ORJpo/G74keNjyjxYO8ezLcZcZz8+00kC4Ra2YS0DWGmPBcR3PUZf/J42kyim8PTGIHryV04gYDSgjuV7hDnEhQn4MBTH9iiBYi2L0hpymUsxgwlYjBYtkJ/HP8Gu5bs+ZRRmS4CXjw7AgwYslzCXuIxQj9B3JwDzoRazoBeLBNAz+48J9NVB7aCRm2UiTsu732Q60ZkpXjtVjOtGpJOEA6n1wTR6FAZHeV5EN+rboEYsZ0jGJ/2DvTdL/wBFCfz1gJvhNEnu5Er0D04bdfB/FZboNXcdeGknfAkuZSGNsX16XV5hjBt1+FnA7keb8tm0k/lf

pjcozzi commented:
@kring any chance you can take a quick look at this?

tfili commented:
@bagnell Yes. I'll open a PR for it once we verify the format is what we need.

bagnell commented:
This looks good to me. Are you going to review again @kring?

kring commented:
Yeah I'd like to. Give me 24 hours?

tfili commented:
@pjcozzi I've made your requested changes. Can you approve your review?

kring commented:
@tfili mostly unrelated to this PR, but do you guys have a GEE server I can use to try the fill-tiles branch with `GoogleEarthEnterpriseTerrainProvider`?

kring commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -840,9 +987,124 @@ define([
+                var key = level + '-' + x + '-' + y;
+                var promise = layer.availabilityPromiseCache[key];
+                if (defined(promise)) {
+                    return promise;
```
It's good there's a cache of promises for tiles loaded for availability purposes, but I think it's employed a bit late. If I'm reading this right, it's still going to create a hella-long promise chain for each position given to `sampleTerrainMostDetailed`. Most of those promise chains won't lead to an actual load because of the cache, but 
when.js promises are still pretty expensive even without a load.

kring replied:
IMO, the `sampleTerrainMostDetailed` algorithm should be something more like this:

1. Find the _known available_ tiles that contain our positions.
2. Load them.
3. Check if we've now learned that there are new, relevant available tiles for our terrain sampling.
4. If so, repeat starting with step 1.

I think that makes more sense than assuming that the max level tiles are available, trying to load those, and working up the tree if that fails. It generates less 404s, which take over a second from Ion currently and return no useful information whatsoever. Also, to a first approximation, max level tiles _don't_ exist. If they did (on average), the terrain dataset would be petabytes or something. The availability of the most detailed terrain is quite spotty, assuming otherwise is a bad assumption.

The algorithm above is basically how the rendering should work too, at least in the fill-tiles branch. Missing availability should stop refinement until availability is known. I think creating promises in the course of checking availability is going to cause way too much overhead and we should avoid it. Promises to load tiles are fine, of course.

By the way, how many levels are of availability are you putting in each tile? I reckon with this new format it can be quite high, like 10+, so it won't substantially impact load time to go top-down when loading availability. Am I wrong?

tfili replied:
This sounds good (and way less complicated to implement).

> By the way, how many levels are of availability are you putting in each tile? I reckon with this new format it can be quite high, like 10+, so it won't substantially impact load time to go top-down when loading availability. Am I wrong?

Right now I'm using 10 levels. I think that will be a good balance of keeping tiles relatively small but maximizing how much availability we can see.

kring commented:
> The one issue with your branch is that we are returning false from getTileDataAvailable. My first thought is that we can just return a promise.

I think promises are pretty heavy-weight for that. Make just an enum? Yes/No/Don't know. It's a breaking changed, but that's true even if we return a promise.

pjcozzi commented:
@kring thanks for the review!

> @pjcozzi I've made your requested changes. Can you approve your review?

I'd like to take another quick look after Kevin's comments are addressed.  I can do it this afternoon if it is ready.

In the meantime, did you notice any Cesium load time improvement as we discussed offline awhile back?

Does this need to be profiled?  Do you have any performance concerns at this point?

tfili commented:
> In the meantime, did you notice any Cesium load time improvement as we discussed offline awhile back?

The only time we save is the download, decompression and parsing of the layer.json. The biggest tileset we have is CWT which currently has the size of 200KB gzipped and 1.8MB uncompressed. The time improvement isn't very noticeable on my laptop.  However, if we end up with a layer.json that is 1.2MB gzipped and 15MB uncompressed then the improvement seems to be about a 1-1.5 seconds on load.

> Do you have any performance concerns at this point?

No. It looks like tiles are loading in about the same time as they did previously.

tfili commented:
Thanks @kring. I think I addressed your concerns. Let me know if you have anything else.

@pjcozzi This should be good for to look at if you have time.

pjcozzi commented:
> if we end up with a layer.json that is 1.2MB gzipped and 15MB uncompressed then the improvement seems to be about a 1-1.5 seconds on load.

Nice!!!

pjcozzi commented on CHANGES.md:
```diff
@@ -9,6 +9,7 @@ Change Log
     * Added `Scene.clampToHeightMostDetailed`, an asynchronous version of `Scene.clampToHeight` that uses the maximum level of detail for 3D Tiles.
 * Added `Scene.invertClassificationSupported` for checking if invert classification is supported.
 * Added `computeLineSegmentLineSegmentIntersection` to `Intersections2D`. [#7228](https://github.com/AnalyticalGraphicsInc/Cesium/pull/7228)
+* Added ability to load availability from quantized mesh tiles every few levels instead of upfront in layer.json. This should speed up the load time of large tilesets significantly. [#7196](https://github.com/AnalyticalGraphicsInc/cesium/pull/7196)
```
Would be good to be specific here like the examples in the comment and frame it in what the end user experiences - faster load time, less memory usage since this is progressive.

pjcozzi commented:
@tfili do you know how code coverage is?  Does this need any additional tests?

pjcozzi commented:
Is there a terrain dataset you can point me to (offline if needed) to test with?

pjcozzi commented:
@kring did you need to look at any of this again?  Or can we just merge when ready?

pjcozzi commented:
> Is there a terrain dataset you can point me to (offline if needed) to test with?

Nevermind, GTOPO + Zion from above is perfect.

kring commented:
The `sampleTerrainMostDetailed` changes look great. I still think it's dodgy to change `getTileDataAvailable` to quietly load tiles, though. It's an undocumented side effect and one that can bring an application to its knees by generating heaps of unexpected requests. This function is called often in the render loop, after all. Does it still need to work this way now that `sampleTerrainMostDetailed` doesn't use it?

kring commented:
Looking at the places `getTileDataAvailable` is used (there are only two), I think just returning undefined when availability is not yet know will work fine. I didn't try it though.

tfili commented:
@kring In master is sort of looks that way, because the parent's must be loaded before the child. I'll mess around with it an do some more testing. I definitely over complicated the problem. In your branch with LOD skipping, you'll need to take that into account though.

@pjcozzi I'll run coverage, but with the simplifications I put in today, I think loading a terrain with metadata availability should cover it, but I'll verify.

kring commented:
> In your branch with LOD skipping, you'll need to take that into account though.

Yeah I haven't totally thought through the implications in the `fill-tiles` branch yet, but you can leave that with me.

mramato commented:
@tfili I think I found an issue with sampling terrain when parent terrain are involved (not sure if it happens in master or not).

1. Run the below example
2. Switch to NED 1m
3. Hit the Sample terrain button

The first time you press it, the samples will all be undeground.  The second time it works.  This doesn't not happen if you just use the GTOPO terrain so it might only be an issue when parent terrains are used.

http://localhost:8080/Apps/Sandcastle/index.html#c=tVhrU9u4Gv4r2n6JMw2K7xegnQOBpeEQ0kK63VI6ZxRbsVUcy2vLkHCG/76vL3HixGHamV0+kCC99+sj+n10kZBIoAFNWTa/vEXEdWmaIsHRkmcJYjxCJE2pSO+jkgYPeYQ9OiNZKE4K4gl/oBF6hzp0eRlML1w2ZpfDz89D5ZoN02F0Y7iDoTl8iP/8Y3DpYCD6y7t4yIkevs4v53dn7mK0VNjox/nT1eTTcjT/KsYXQ208UObXk5vg+syDO38xmo/Y1eAyvgNho+UTI19+l4c/+OJ6MlRHZ5+N6+eTp9kn7M4Gn/4SB4F+5+mC/O/P8x/PwrEyzR6QsfO4WDzNrNtPp/Zw6g07R20+3dLkkSa5O4EQcXrY73v08YDEDLslrcvnOed99EgS5Ase8wlNEsLyEET0qQolLj+qq48Jf2QeTaT/30cIfrIkPERr3Tc0hWC7FM8SPj/Jwz30JMvu3kcv3VrVM+TiX9LkNDU9MvpUhGBDyR/FmdQpgzDgkQBlNOn0Kj2iqR4dNkJTi3/mfD7hec15AxIteSQ1HTxfCArl+G6l9oa6gkR+WBp8QzxGolQ6ULBjqZYqy5aiyaqi62pPxqZu2bpty5pmOZYFRwUZkKiaoVqGqVi2VpA5umEYig4npqnYlQGFfzh1aUSxH/IpxTQi05BeMT8QLPLBJpFktCC+j27BAZekAgwjnjfhPJySZESjTPpWB2QhIAqdi8n441iTO73ymEcpDcEpuJplkStYHgJU8eQ/lSHb8XzXiOdRSf8Cce2hLYXX52dImSM+Q3cgvIc+T0ApNHScH/1j1mxU44Yx38tg/kJ08kpAZSn8tFHtRbQ3IKsw/KL8O7YluHJuxYPchBJBLxLmScmqTD+QcHbLnmktrxgSQPKFeSKAuOnKUfPiA80LbPuGwgyiqbgigonMo+uGGBERYMFXnaBa2LFt1bC6Lcw88l/ntk3sqHknNLibjV93oLQj9gDt+N3bsfxVolrU258R1UK0aXhVpLdkHof0I09ZnqUUvPn2vaKa8QRJOekSTuUj+DjeyMERevt22SiEmn5R0i8q+iKZOfmiQb4yJNwT+JAmsUTxE/gEzmFK8s8F6iNpXSEHSOmunGrIbK+ESiQMdRHkMiOe5F+WK6FVde2TGldR2looJBHcT0gcMFeqnenVNmxLag88jrM0kFYaNnle6p7KPxIqsiTaIwS4XhpNF3L+cCJG4rysDanRaIIkPhW7zsA3EmmSBvvCkrEh25aq2Y5m6z1kmJruKCpWNEWDvWAbPaQ6lq0YKlZt0zB0Q1HMzSrjs1n6ig4Qp8PW0WXHNi3bMNUeOrAcDWu64liOLduKpSugQ3cc7IAltilbpqaYaq2jnLgumdOE4NJdqXSsVyl/jXQCUzGFup1L6zJJ2ELHw7NzgEqTr91VSFuH9GkmBIy9TpkGNOKpQGcUdn1IPVTFHFVzv9NrmZ9shqTfKtWAqAAjeFL7EsHkEcSSKQuZWHYbjeRC5jnYFXJf6qSFKZXO3KDaHpaiiAuUZnEMZQ8Heb+KgKJyvsNBpRLFlc7OZhmWhXfUKMZX58jGwJexLMv12Kz8fQpotAr7Xqv3RKO3R223hxqibrMCc68077TDUQ216pZp45f2aNvsJhqGLE4589Yz53x1hL9c3Np6oScn7/dPl6gC0L2KGHmclgni0zRLKPIpn1ORLFcsUxqwqM4RRrdUFEBLBJDYWUh8VCIweI0EROBGzW/iNI/GIpiA9yc+yEnFGiTXgG2DE+Al+EpTnGZpTCMPIheJVNrqqZoqoXP+SE/CsAKqjbXAyrXAYC3smYEhjfxyVbAmrmoO33bub+x7rbPNNmhbaWv9RDALANesROOgmP+w8n9nC6g9pdtr0tc2HK7TDQNlPf8nvB5u61m+JQRaOJxyknggZcuewmyaCOaScJwwn+WaVk+Kxjk+HU8m41Fvlz8FmtwpGVstt2xO/Py2g3G/+J72Z8Qthgr22azT5HjZkgDzh4btZlfw8adDWVZGVEBORY4FmvOIpzFxaaeFMuAJe85fUTth+bB1gwcwuM9vWmTEYEg4LvfRYes+UiUZ9o+it9oKWRvwkCdr1cWf+PTqZPDfFgbAGCHM81/iSQP+dErcBz/hWZTXR96SLXTTmmYlftOd/AhGrtNDm7+s7quCPkJ75ONkX2Rg+WvdrerYwCjdrcWwMxfSbE43hkcTprQ8UerR2tjas3A54Ztd7MEgYxGpmnIHYBTv4DPqJ5Tm7+Acuehm/t41NcNRVROcMrFsaLasqjpgEcM04QweCDLWDEuDt7Am67ZmbsYOCg7cWKncaoWAkiqMe14gDjbhEWEphq2bjmIZOyOGCTfYy36gWdiwTYBemqkAHNK1bf6Eh2HR/Qo8VBx47pu2o1q6aRu7iHJfIu7Yv5kBxzYcQwMcp9qampeVBUkBTGeoYCzABMCUuq0b2IYMyLaRu6D+c/FXsS5bugXQFXKrQlx+NQHwCLRlXbV1VYeXpKkaezMA+QHIKoMvsuqAz/YrGbiP3vTeHKdiGdL35d1/2DzHafm/pCQY1oLCsgM8lfanmftABXZLWHPcXzEde+wRMe/d/Zut/zrdv0FuSNIUbmZZGOYPwfs374/7QN9gC3kRuzFsoJAsgaQ05DhQ3l+VVxjj4z78mavd5hYlJN6Q/Dc

tfili commented:
@kring returning undefined seems to work well. The one caveat is that if we are using parent terrain, I still need to do a fetch tile for the parent during `getTileDataAvailable` because we will never load a level 0 tile because the top level terrain will always have it. The single layer common case will do none of that, so I think it's ok. 

@mramato I'm looking at your issue now.

mramato commented on Source/Core/TerrainProvider.js:
```diff
@@ -227,5 +227,15 @@ define([
      */
     TerrainProvider.prototype.getTileDataAvailable = DeveloperError.throwInstantiationError;
 
+        /**
```
Formatting

mramato commented on Source/Core/TerrainProvider.js:
```diff
@@ -227,5 +227,15 @@ define([
+     * @param {Number} level The level of the tile for which to request geometry.
+     * @returns {undefined|Promise} Undefined if nothing need to be loaded or a Promise that resolves when all required tiles are loaded
+     */
+    TerrainProvider.prototype.loadTileDataAvailability = DeveloperError.throwInstantiationError;
```
This is a breaking change because existing terrain providers people have written will need to be updated.  While that's unlikely to affected most (all?) of our users, we should list it as such in CHANGES

mramato replied:
Basically, if they implement availability, they need to implement loadTileDataAvailability, right?

mramato commented:
Just those last two comments.

mramato commented on CHANGES.md:
```diff
@@ -3,13 +3,17 @@ Change Log
 ### 1.52 - 2018-12-03
 
+##### Breaking Changes :mega:
+* `TerrainProviders` must now implement the `loadTileDataAvailability` method.
```
```suggestion
* `TerrainProviders` that implement `availability` must now also implement the `loadTileDataAvailability` method.
```

mramato commented:
Thanks @tfili.

@kring I'm going to merge this so it has some time to stew in master, but if you run into any issues in the `fill-tiles` branch or if you have any additional feedback, let us know and we'll get it in ASAP. (Same goes for everyone else for that matter).
